Convex Avatar adaptations

petrovska-petro opened the pull request:
tackles #17

shuklaayush commented on src/convex/ConvexAvatarUtils.sol:
```diff
@@ -0,0 +1,57 @@
+        usdcAmount_ = (_cvxAmount * cvxInEth) / FEED_DIVISOR_ETH;
+    }
+
+    function fetchPriceFromClFeed(IAggregatorV3 _feed) internal view returns (uint256 answerUint256_) {
+        (, int256 answer,, uint256 updateTime,) = _feed.latestRoundData();
+
+        if (block.timestamp - updateTime > CL_FEED_HEARTBEAT) {
+            revert StalePriceFeed(block.timestamp, updateTime, CL_FEED_HEARTBEAT);
+        }
+
+        answerUint256_ = uint256(answer);
+    }
```
Let's move this to a `BaseAvatarUtils` contract

petrovska-petro replied:
pushed that contract with the cl utils in 2d5c0dd5355515d9cc47c3acddc4f7fd02a97486

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,205 @@
+            1, 0, _cvxAmount, (getCvxAmountInEth(_cvxAmount) * minOutBpsCrvToUsdc.val) / MAX_BPS
+        );
+
+        // 2. Swap cvx -> usdc
```
```suggestion
        // 2. Swap eth -> usdc
```

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,205 @@
+    function swapCrvForUsdc(uint256 _crvAmount) internal returns (uint256 usdcEarned_) {
+        uint256 minUsdcExpected = (getCrvAmountInUsdc(_crvAmount) * minOutBpsCrvToUsdc.val) / MAX_BPS;
+        // 1. Swap crv -> eth
+        CRV_ETH_CURVE_POOL.exchange_underlying(
+            1, 0, _crvAmount, (getCrvAmountInEth(_crvAmount) * minOutBpsCrvToUsdc.val) / MAX_BPS
+        );
```
Why use ETH for the swaps instead of WETH?

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,205 @@
+            1, 0, _crvAmount, (getCrvAmountInEth(_crvAmount) * minOutBpsCrvToUsdc.val) / MAX_BPS
+        );
+
+        // 2. Swap eth -> usdc
+        uint256 ethBal = address(this).balance;
+        IUniswapRouterV3.ExactInputParams memory params = IUniswapRouterV3.ExactInputParams({
+            path: abi.encodePacked(WETH, uint24(500), address(USDC)),
+            recipient: address(this),
+            amountIn: ethBal,
+            amountOutMinimum: minUsdcExpected
+        });
+        usdcEarned_ = UNIV3_ROUTER.exactInput{value: ethBal}(params);
+        // NOTE: conservative approach given https://hackmd.io/@0x534154/Hy6_66gT5
+        UNIV3_ROUTER.refundETH();
```
Let's combine the ETH --> USDC swaps of both into a single swap

petrovska-petro replied:
combined in 2d5c0dd5355515d9cc47c3acddc4f7fd02a97486

shuklaayush reviewed (changes requested):
Left some comments

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+        address _owner,
+        address _manager,
+        address _keeper,
+        uint256[] memory _pids,
+        uint256[] memory _fraxPids
```
```suggestion
        uint256[] calldata _pids,
        uint256[] calldata _fraxPids
```

petrovska-petro replied:
resolved in 8f1285e8799cd5cb413b8cb597f1829926afab18

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+    // PUBLIC: Manager - Config
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Updates the current value for the minimum executable price (in bps as proportion of an oracle price)
+    ///         for a CRV to WETH swap. The value should be more than the minimum value. Can be called by the owner or
+    ///         the manager.
+    /// @param _minOutBpsCrvToWeth The new value in bps.
+    function setMinOutBpsCrvToWethVal(uint256 _minOutBpsCrvToWeth) external onlyOwnerOrManager {
+        if (_minOutBpsCrvToWeth > MAX_BPS) {
+            revert InvalidBps(_minOutBpsCrvToWeth);
+        }
+
+        uint256 minOutBpsCrvToWethMin = minOutBpsCrvToWeth.min;
+        if (_minOutBpsCrvToWeth < minOutBpsCrvToWethMin) {
+            revert LessThanBpsMin(_minOutBpsCrvToWeth, minOutBpsCrvToWethMin);
+        }
+
+        uint256 oldMinOutBpsCrvToWethVal = minOutBpsCrvToWeth.val;
+        minOutBpsCrvToWeth.val = _minOutBpsCrvToWeth;
+
+        emit MinOutBpsCrvToWethValUpdated(_minOutBpsCrvToWeth, oldMinOutBpsCrvToWethVal);
+    }
```
Do we need a similar function for other bps'?

petrovska-petro replied:
i decided just to add one in the test, since all of them contains the same structure, only diff is the naming and purpose. but test would be identical, i think may be redundant

petrovska-petro replied:
initially i misunderstood the question, thinking that was referring to the test file, and replied that with one was enough, since otherwise will be testing the same functionalities. 

For the rest of variables is required to have setters, yes. I though of the approach of the mapping initially, but got bit ugly in the code. 

Something like reading this in swap methods:

```
bpsTkn2TknConfig[address(FXS)][address(FRAX)].val

or 

bpsTkn2TknConfig[address(WETH)][address(DAI)].val
```

But also is true that it can get ugly on the setters with many of them doing the same. Not strong opinion in either approach tbh

petrovska-petro replied:
mapping architecture introduced in 084a9d54d5f9a8f537e79730245ab6c0f9b58d42

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+        }
+
+        IStakingProxy proxy = IStakingProxy(vaultAddr);
+        address stakingToken = proxy.stakingToken();
+        IFraxUnifiedFarm farm = IFraxUnifiedFarm(proxy.stakingAddress());
```
Should we cache `stakingToken` and `stakingAddress` to storage?

petrovska-petro replied:
in this case was only used once, not sure if worth caching here

always we can derived those two vars from the proxy itself which is on storage, for the convex-frax infra i decided only to store the keks and proxies

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+        bytes32[] memory keks = kekIds[vaultAddr];
+
+        for (uint256 i = 0; i < keks.length;) {
+            proxy.withdrawLockedAndUnwrap(keks[i]);
```
Shouldn't we also add a function to withdraw from a subset of keks since this will fail if not all of them have unlocked?

petrovska-petro replied:
that's a good point, let me think about how to handle perhaps in the loop all those which are unlocked and another method to pinpoint to specific keks if required

petrovska-petro replied:
resolved in 3aa143db86fd36de76d5c5b0f57272d256e6207f

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+    /// @dev This function doesn't claim any rewards.
+    function withdrawAll() external onlyOwner {
+        uint256 length = baseRewardPools.length();
+        uint256[] memory bptsDeposited = new uint256[](length);
```
Rename bpt to something else

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+        emit RewardClaimed(address(FXS), totalFxs_, block.timestamp);
+    }
+
+    function swapCrvForWeth(uint256 _crvAmount) internal {
+        // Swap CRV -> WETH
+        CRV_ETH_CURVE_POOL.exchange(
+            1, 0, _crvAmount, (getCrvAmountInEth(_crvAmount) * minOutBpsCrvToWeth.val) / MAX_BPS
+        );
+    }
```
Let's return the weth amount here and use that instead of doing `WETH.balanceOf`

petrovska-petro replied:
returned and use it then for the other method `weth` for `dai` in 8f1285e8799cd5cb413b8cb597f1829926afab18

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+        uint256 fxsPending;
+        uint256 length = baseRewardPools.length();
+
+        for (uint256 i = 0; i < length;) {
```
```suggestion
        for (uint256 i; i < length;) {
```

petrovska-petro replied:
added the suggestion across the contract 8f1285e8799cd5cb413b8cb597f1829926afab18

shuklaayush commented on src/convex/ConvexAvatarUtils.sol:
```diff
@@ -0,0 +1,43 @@
+    ////////////////////////////////////////////////////////////////////////////
+    // INTERNAL VIEW
+    ////////////////////////////////////////////////////////////////////////////
+    function getFxsAmountInFrax(uint256 _fxsAmount) internal view returns (uint256 usdcAmount_) {
+        uint256 fxsInUsd = fetchPriceFromClFeed(FXS_USD_FEED, CL_FEED_DAY_HEARTBEAT);
+        // Divisor is 10^8 and uint256 max ~ 10^77 so this shouldn't overflow for normal amounts
+        usdcAmount_ = (_fxsAmount * fxsInUsd) / FEED_DIVISOR_FXS_USD;
+    }
```
Maybe add a note here that this is assuming frax:usd is 1:1

petrovska-petro replied:
comment added in the assumption for that method in 8f1285e8799cd5cb413b8cb597f1829926afab18

shuklaayush commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+    /// @dev This function doesn't claim any rewards. Caller can only be owner.
+    /// @param _pids Pids to be targetted to unstake from.
+    /// @param _bptsDeposited Amount of assets to be unstaked.
+    function _withdraw(uint256[] memory _pids, uint256[] memory _bptsDeposited) internal {
```
```suggestion
    function _withdraw(uint256[] calldata _pids, uint256[] calldata _bptsDeposited) internal {
```

petrovska-petro replied:
resolved in 8f1285e8799cd5cb413b8cb597f1829926afab18

petrovska-petro commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,698 @@
+    /// @notice Takes a given amount of assets from the owner and stakes them on the CONVEX Booster. Can only be called by owner.
+    /// @param _pids Pids target to stake into
+    /// @param _amountAssets Amount of assets to be staked.
+    function deposit(uint256[] memory _pids, uint256[] memory _amountAssets) external onlyOwner {
```
resolved in 8f1285e8799cd5cb413b8cb597f1829926afab18

gosuto-inzasheru commented:
note: did not review tests yet, neither code after L282

petrovska-petro commented on src/convex/ConvexAvatarUtils.sol:
```diff
@@ -0,0 +1,46 @@
+    // INTERNAL VIEW
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @dev Assumes frax:usd is 1:1
```
remove ratio assumption in c58b1841140ca671f88c320e05b22bc3c41a4119

petrovska-petro commented on src/BaseConstants.sol:
```diff
@@ -12,3 +12,7 @@ address constant BADGER_PROXY_ADMIN = 0x20Dce41Acca85E8222D6861Aa6D23B6C941777bF
+
+// Convex
+uint256 constant CONVEX_PID_BADGER_WBTC = 74;
+uint256 constant CONVEX_PID_BADGER_FRAXBP = 35;
```
at the end similar like i did in aura avatar i moved it to the test file itself to be isolated, see [41a35a9aae11cc19a075c00a364ad30a3cc247eb](https://github.com/Badger-Finance/badger-avatars/pull/19/commits/41a35a9aae11cc19a075c00a364ad30a3cc247eb#diff-de0cefae94692b8b4f69ec4e23a8610f0d8aa4a01f958767c9342770ace5ee4f)

gosuto-inzasheru commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -119,10 +107,8 @@ contract ConvexAvatarMultiToken is BaseAvatar, ConvexAvatarUtils, PausableUpgrad
-    event MinOutBpsCvxToWethValUpdated(uint256 newValue, uint256 oldValue);
-    event MinOutBpsFxsToFraxValUpdated(uint256 newValue, uint256 oldValue);
-    event MinOutBpsWethToDaiValUpdated(uint256 newValue, uint256 oldValue);
+    event MinOutBpsValUpdated(address tokenA, address tokenB, uint256 newValue, uint256 oldValue);
+    event MinOutBpsMinUpdated(address tokenA, address tokenB, uint256 oldValue, uint256 newValue);
```
why `newValue, oldValue` in one and `oldValue, newValue` in the other?

gosuto-inzasheru replied:
looks like you are using it as new,old in L256

petrovska-petro replied:
good catch

gosuto-inzasheru commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -57,21 +57,9 @@ contract ConvexAvatarMultiToken is BaseAvatar, ConvexAvatarUtils, PausableUpgrad
-    BpsConfig public minOutBpsFraxToDai;
+    /// @notice The current and minimum value (in bps) controlling the minimum executable price (as proportion of oracle
+    ///         price) for a tokenA to tokenB swap.
+    mapping(address => mapping(address => BpsConfig)) public bpsTkn2TknConfig;
```
why not just call this `minOutBps`?

gosuto-inzasheru commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -327,14 +328,22 @@ contract ConvexAvatarMultiToken is BaseAvatar, ConvexAvatarUtils, PausableUpgrad
         bytes32 kek = kekIds[vaultAddr];
 
-        // TODO: double-check that indeed that kek_id is expired
+        /// @dev: only inspecting index `0` given that we only use an unique lock id per private vault
+        (,,, uint256 endingTimestamp,) = unifiedFarm.lockedStakes(vaultAddr, 0);
```
is this sufficient in case a second deposit is initiated after a kek_id has expired and been replaced with a new deposit? ie
- deposit (kek_id1)
- withdraw after lock expires
- deposit again (new id, kek_id2, no?)
- check if expired lock will now check `endingTimestamp` of kek_id1 no?

gosuto-inzasheru replied:
can confirm by asserting `kekId == unifiedFarm.lockedStakes(vaultAddr, 0)[0]`

petrovska-petro replied:
my understanding is that when the 2nd action happens wd after lock expires is that the kek_id also within the unified farm contract will be delated from their storage, [ref](https://etherscan.io/address/0x5a92ef27f4baa7c766aee6d751f754ebdebd9fae#code#L2137), so then we can deposit again, it will create lock in the index 0, since the avatar will not handle two locks in tandem, i think checking for zero index seems unless i am missing something

gosuto-inzasheru reviewed (commented):
some misplaced `///` for comments, global regex search for `/// [^@ ]`

GalloDaSballo commented:
Notes:
https://mint-salesman-909.notion.site/Convex-Avatar-313e7fdb46a2493b9ecda048c787e363

gosuto-inzasheru reviewed (commented):
minor stuff. did not review after L376 yet

petrovska-petro commented:
> some misplaced `///` for comments, global regex search for `/// [^@ ]`

see commit c78b60726e6014b0f442fb890aeb9015b02c1457, i think all were sorted

petrovska-petro commented:
> feat: limit the risk of misconfigured min locking time from the unifa…

added suggestion of checking for the min lock time in 81015c161ad4b14c3c096ede727c0c98f1c67bdb and added test around malicious env in 8825b92ed71b43359cefbd6208d072184dbf7921

refactor the tracking of rewards in the avatar to have tracking per block, see on 1ab3d159b0323f198d6d4b3871dafae9f52735d0

petrovska-petro commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,816 @@
+    // PUBLIC: Owner
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Takes a given amount of asset from the owner and stakes them on private vault. Can only be called by owner.
```
sorted in ca13862dd2a32dd477e88747532de8e3e0f06d9f

petrovska-petro commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,816 @@
+        emit Deposit(stakingToken, _amountAsset, block.timestamp);
+    }
+
+    /// @notice Unstakes all staked assets and transfers them back to owner. Can only be called by owner.
```
sorted in ca13862dd2a32dd477e88747532de8e3e0f06d9f

petrovska-petro commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,816 @@
+    }
+
+    /// @notice Unstakes all staked assets and transfers them back to owner. Can only be called by owner.
+    /// It will loop thru all existent keks and withdraw fully from each of them.
```
sorted in ca13862dd2a32dd477e88747532de8e3e0f06d9f

GalloDaSballo commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,833 @@
+        uint256 curveLpBalance = curveLp.balanceOf(address(this));
+        curveLp.safeTransfer(owner(), curveLpBalance);
+
+        delete kekIds[vaultAddr];
```
Per discussions we've already had, this is not CEI conformant (change storage after external interactions)

This function is callable only by owner / manager so it's fine.

But you should fix it to avoid making the mistake when stakes are higher, recommend you read from storage (`kek`) then just delete the value from storage right after (before the external calls)

petrovska-petro replied:
added suggestion in  c376d84c41c74160ef5d845e12eb9bc06a50ac0c

GalloDaSballo commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,833 @@
+    // INTERNAL
+    ////////////////////////////////////////////////////////////////////////////
+
+    function _createPrivateVault(uint256 _pid) internal {
```
Since we could be calling this multiple times (we could create 2 vaults with same pid)

Is there ever a reason to remove the allowance we are granting?

`IERC20MetadataUpgradeable(stakingTokenAddr).safeApprove(vaultAddr, type(uint256).max);`

petrovska-petro replied:
in this particular case, do not see a concern leaving the allowance granted to the private vault, since concerning methods are permissioned to only `owner` (in this case the creator of the vault, the avatar) to trigger methods which could move funds, such as `stakeLockedCurveLp` and alike

GalloDaSballo commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,832 @@
+        }
+
+        // NOTE: verify pending rewards and claim. Processing is done separately
+        if (baseRewardPool.earned(address(this)) > 0) {
```
I wonder if you can getReward if rewards are zero, I think it would save gas

petrovska-petro replied:
looks safe to call it even when rewards are zero

https://etherscan.io/address/0x36c7E7F9031647A74687ce46A8e16BcEA84f3865#code#L980

internal call to one virtual base reward pool containing extra rewards 

https://etherscan.io/address/0x008aEa5036b819B4FEAEd10b2190FBb3954981E8#code#L874

it will not revert 👍

petrovska-petro replied:
resolved in f1439b3604993b3d17daaf60f95e55ad709718ef

GalloDaSballo commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,832 @@
+            IUniswapRouterV3.ExactInputSingleParams({
+                tokenIn: address(WETH),
+                tokenOut: address(DAI),
+                fee: uint24(500),
```
FUTURE NOTE: Pricer + SwapExec Opportunity

GalloDaSballo commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,832 @@
+            }
+            kekId = proxy.stakeLocked(_amountAsset, farm.lock_time_min());
+            /// @dev detailed required to enable withdrawls later
+            kekIds[vaultAddr] = kekId;
```
Same issue with CEI

GalloDaSballo replied:
This may be the one time you must use a nonReentrant modifier (if necessary obv)

```
            proxy.lockAdditional(kekId, _amountAsset);
        } else {
            // NOTE: we always try to lock for the min duration allowed
            uint256 minLockTime = farm.lock_time_min();
            if (minLockTime > MAX_LOCKING_TIME) {
                revert MinLockTimeMisconfigured(address(farm), minLockTime);
            }
            kekId = proxy.stakeLocked(_amountAsset, farm.lock_time_min());
            /// @dev detailed required to enable withdrawls later
            kekIds[vaultAddr] = kekId;
        }
```

petrovska-petro replied:
considerations while double checking reentrancy concern:

- method itself is permissioned to the owner, in this case the avatar
- contract does have already in [stakeLocked](https://github.com/convex-eth/frax-cvx-platform/blob/main/contracts/contracts/StakingProxyConvex.sol#L92) a `nonReentrant` modifier
- does not present fallback methods
- approvals from avatar's owner will be limited to the `_amountAsset` per atomic tx ideally

re CEI as discussed in dms this particular case, it is not possible to be cei compliant due to the dependancy of the `kekId` from the external call and after we write into storage

will drop a commit with minor comments to avoid other falling into the same concern in the same L

petrovska-petro replied:
comments on DD in c908449d96a23e6a9808edceff7116d773790a2d

GalloDaSballo reviewed (commented):
Mostly LGTM, sent minor notes.

I think you should look into pricer and figure out if worth getting V4 out to support work like this

GalloDaSballo reviewed (approved):
LGTM, Let's road test it

GalloDaSballo commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,848 @@
+            IERC20MetadataUpgradeable(lpToken).safeTransfer(owner(), _curveLpDeposited[i]);
+
+            emit Withdraw(lpToken, _curveLpDeposited[i], block.timestamp);
+            unchecked {
+                ++i;
+            }
```
Saves gas

GalloDaSballo reviewed (dismissed):
Uses EnumerableSet to skip a call to BOOSTER

All tests passing

LGTM

petrovska-petro commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,848 @@
+            address lpToken = assets.at(pids.indexOf(_pids[i]));
+            IERC20MetadataUpgradeable(lpToken).safeTransferFrom(msg.sender, address(this), _amountAssets[i]);
+
+            CONVEX_BOOSTER.deposit(_pids[i], _amountAssets[i], true);
```
intention was to keep methods naming generalised `deposit`, `withdraw`, `pendingRewards` and alike. with the goal of ending up with a generalised interface for the avatars in the future. similar to the strategies templates 

regardless, if within the deposit method we are taking a leverage position in x/y protocol or in the withdrawal method deleveraging from those same protocol for example

but if there is value added in being + explicit in internal actions, we can update it. but wanted to provide a bit of the reasoning/background

gosuto-inzasheru replied:
i can see that, just maybe good to add some comments then to make it clear that there is a deposit and stake happening in this line.

petrovska-petro replied:
added more comments in 3aea5c486f527ebe1ebc02e79d835678b094a5a0

petrovska-petro commented on src/convex/ConvexAvatarMultiToken.sol:
```diff
@@ -0,0 +1,848 @@
+    }
+
+    /// @notice Unstakes the given amount of assets and transfers them back to owner.
+    /// @dev This function doesn't claim any rewards. Caller can only be owner.
```
good catch, i think given the external method comment guidance, this becomes redundant, will trim it out in another commit

petrovska-petro replied:
trimmed out in 3aea5c486f527ebe1ebc02e79d835678b094a5a0

gosuto-inzasheru reviewed (approved):
green lighting, but will let you be the judge of whether that means this has gotten sufficient solidity review or not

GalloDaSballo reviewed (approved):
Just comments, LGTM

Per discussions, let's cap at 100k for first couple weeks, and trial it
